hypothesis-python/src/hypothesis/core.py

```diff
@@ -1248,6 +1250,22 @@ def wrapped_test(*arguments, **kwargs):
                     "to ensure that each example is run in a separate "
                     "database transaction."
                 )
+            if settings.database is not None:
+                nonlocal prev_self
+                cur_self = (
+                    stuff.selfy
+                    if getattr(type(stuff.selfy), test.__name__, None) is wrapped_test
```

Comment:
Not beautiful, but we really need to be reasonably sure that `selfy` is really `self` at this point, to avoid false positives.

We could move this test, or maybe something better, into `process_arguments_to_given`.

Reply:
Yes, and I'll take to heart that when I feel the need for pointing something out for review it should probably be a code comment. In this instance no action from me because you already did, right?